Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU Library General Public License
+ * along with this library; see the file COPYING.LIB.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "config.h"
+#include "RenderSVGResourcePattern.h"
+
+#if ENABLE(LAYER_BASED_SVG_ENGINE)
+#include "ElementChildIteratorInlines.h"
+#include "RenderSVGResourcePatternInlines.h"
+#include "RenderSVGShape.h"
+#include "SVGFitToViewBox.h"
+#include "SVGRenderStyle.h"
+#include "SVGRenderingContext.h"
+#include <wtf/IsoMallocInlines.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGResourcePattern);
+
+RenderSVGResourcePattern::RenderSVGResourcePattern(SVGElement& element, RenderStyle&& style)
+    : RenderSVGResourcePaintServer(Type::SVGResourcePattern, element, WTFMove(style))
+{
+}
+
+RenderSVGResourcePattern::~RenderSVGResourcePattern() = default;
+
+void RenderSVGResourcePattern::collectPatternAttributesIfNeeded()
+{
+    if (m_attributes.has_value())
+        return;
+
+    auto attributes = PatternAttributes { };
+
+    SVGPatternElement* current = &patternElement();
+
+    patternElement().synchronizeAllAttributes();
+
+    while (current) {
+        current->collectPatternAttributes(attributes);
+
+        auto target = SVGURIReference::targetElementFromIRIString(current->href(), current->treeScopeForSVGReferences());
```

Comment:
This is still a concern and source of confusion- we should move more logic into collectPatternAttrs as for gradients...

Reply:
I think patterns have a different kind of recursion which is hard to combine with collectFoo, i.e. the pattern content can reference other patterns, this can't happen with gradients. But yes, can be visited in the future.